Handle `descriptor` field from `FragSessionSetupReq` at application level

sophiekovalevsky opened the pull request:
The following pull request proposes a logic to add a callback into the `frag_transport` module so that , at application level, the fuota process can be aborted if `Descriptor` field sent from `FragSessionSetupReq` does not meet the application conditions to keep going. 

Here is an excerpt of what the Fragmented Data Block Transport [TS004-1.0.0](https://resources.lora-alliance.org/technical-specifications/lorawan-fragmented-data-block-transport-specification-v1-0-0) stated about this field:
![image](https://github.com/user-attachments/assets/129924af-5ff2-4434-a23c-953f07b2cf76)


The logic implemented has been similar to the one used in the module `lorawan` when external callbacks are added for example to obtain the battery level of devices. 

Also, an extra piece of code has been added in lorawan fuota sample to illustrate how to use it.

sophiekovalevsky committed:
lorawan: add callback for descriptor changes

allow the application to decide whether to keep going with the fuota
process by setting a callback that exposes the descriptor field
whenever `FragSessionSetupReq` is sent to the device.

Signed-off-by: Kiara Navarro <sophiekovalevsky@fedoraproject.org>

sophiekovalevsky commented:
@JordanYates, @martinjaeger, @kartben  may you please check this?

sophiekovalevsky commented:
Hi, I would like if somebody can take a look at this. I would like the label "stale" to be removed.

martinjaeger commented:
Sorry for the delay with the review. Somehow missed it on my todo list. I removed the Stale label and will take a look ASAP.

sophiekovalevsky commented:
@martinjaeger thanks so much for taking care of this. Let me know if you need some rebase here for you to test, the issue has been inactive for more than one month and I think it was created over the time of releasing new zephyr version.

martinjaeger commented on include/zephyr/lorawan/lorawan.h:
```diff
@@ -406,6 +406,20 @@ int lorawan_clock_sync_get(uint32_t *gps_time);
+ *
+ * @return 0 if successful, negative errno code if failure
+ */
+void lorawan_frag_transport_register_handle_descriptor_callback(
+	int (*transport_handle_descriptor)(uint32_t descriptor));
```
Can we name this a bit more simple?

```suggestion
void lorawan_frag_transport_register_descriptor_callback(int (*descriptor_cb)(uint32_t descriptor));
```

sophiekovalevsky replied:
I can not change the argument for `descritor_cb` since will overlap when assigning the callback.

martinjaeger commented on samples/subsys/lorawan/fuota/src/main.c:
```diff
@@ -42,6 +42,16 @@ static void datarate_changed(enum lorawan_datarate dr)
 
+int handle_descriptor(uint32_t descriptor)
+{
+	/*
+	 * In an actual application the firmware may be able to handle
+	 * the descriptor field
+	 */
```
Why not print it out with `LOG_INF`?

martinjaeger replied:
I meant printing out the actual descriptor, not the comment.

martinjaeger commented on subsys/lorawan/services/frag_transport.c:
```diff
@@ -101,6 +101,9 @@ static struct frag_transport_context ctx;
 static void (*finished_cb)(void);
 
+/* Callback to handle descriptor field */
+static int (*handle_descriptor)(uint32_t descriptor);
```
For consistency with the callback above:

```suggestion
static int (*descriptor_cb)(uint32_t descriptor);
```

martinjaeger replied:
Looks like you accidentally dropped this line, which is why twister is now failing.

sophiekovalevsky replied:
Thanks for this. While I was doing the pushed, just noticed that I needed to do a rebase and in that process forgot to rebuild again.

martinjaeger reviewed (commented):
Sorry for the late review. Only got some minor suggestions for improvement.

martinjaeger reviewed (changes requested):
Looks good now.

Only missing thing: Please squash all changes into one single commit, as we don't accept fix-up commits in PRs in Zephyr.

martinjaeger commented:
> Should only be a single commit? I think we should separate the implementation from the example. 

Yeah, you are right. 2 commits are perfect.

JordanYates commented on include/zephyr/lorawan/lorawan.h:
```diff
@@ -439,6 +439,20 @@ int lorawan_clock_sync_get(uint32_t *gps_time);
+ *
+ * @param transport_descriptor_cb Callback for notification.
+ *
+ * @return 0 if successful, negative errno code if failure
```
Function returns `void`, not an `int`

JordanYates reviewed (changes requested):
Feature and implementation look fine. I would just like the callback function to be typedef'd so that it can be documented properly. e.g. something like:
```
/**
 * @param descriptor Arbitrary 4 byte payload which describes the data block.
 *                  refer to network server for interpretation details
 * @retval 0 when session should continue
 * @retval -errno if session should be aborted
 */
typedef int (*transport_descriptor_cb)(uint32_t descriptor);
```

sophiekovalevsky committed:
samples: subsys: lorawan: callback registration for handle descriptor

include the use of callback registration to handle descriptor
whenever `FragSessionSetupReq` is sent from fuota server.

Signed-off-by: Kiara Navarro <sophiekovalevsky@fedoraproject.org>

sophiekovalevsky commented:
> Feature and implementation look fine. I would just like the callback function to be typedef'd so that it can be documented properly. e.g. something like:
> 
> ```
> /**
>  * @param descriptor Arbitrary 4 byte payload which describes the data block.
>  *                  refer to network server for interpretation details
>  * @retval 0 when session should continue
>  * @retval -errno if session should be aborted
>  */
> typedef int (*transport_descriptor_cb)(uint32_t descriptor);
> ```

I just took the example of how the signature of `lorawan_frag_transport_run` was. Maybe this is something that I can standardize so that all signatures that are registering callbacks are properly typedefs?  I would be more than happy to help on this change.

JordanYates commented on include/zephyr/lorawan/lorawan.h:
```diff
@@ -217,6 +217,15 @@ typedef uint8_t (*lorawan_battery_level_cb_t)(void);
+/**
+ * @brief Defines the descriptor callback handler function signature.
+ *
+ * @param descriptor Descriptor field
```
Where does this value come from, what does it mean, and how should the user use the value?
This is a function the user needs to implement, so the docs need to explain what the expected inputs & outputs are.

JordanYates commented on include/zephyr/lorawan/lorawan.h:
```diff
@@ -217,6 +217,15 @@ typedef uint8_t (*lorawan_battery_level_cb_t)(void);
+ *
+ * @param descriptor Descriptor field
+ *
+ * @return 0 if successful, negative errno code if failure
```
The user is implementing this, so it should document what values should be returned in different circumstances.
What value do you return to proceed with the download, what value to terminate the download?

JordanYates reviewed (changes requested):
`typedef` docs need to be less generic.
